test/unit_tests/wiring/variant.cpp

```diff
@@ -0,0 +1,474 @@
+    if constexpr (std::is_same_v<T, bool>) {
+        CHECK(v.asBool() == expectedValue);
+    } else if constexpr (std::is_same_v<T, int>) {
+        CHECK(v.asInt() == expectedValue);
+    } else if constexpr (std::is_same_v<T, unsigned>) {
+        CHECK(v.asUInt() == expectedValue);
+    } else if constexpr (std::is_same_v<T, int64_t>) {
+        CHECK(v.asInt64() == expectedValue);
+    } else if constexpr (std::is_same_v<T, uint64_t>) {
+        CHECK(v.asUInt64() == expectedValue);
+    } else if constexpr (std::is_same_v<T, double>) {
+        CHECK(v.asDouble() == expectedValue);
+    } else if constexpr (std::is_same_v<T, String>) {
+        CHECK(v.asString() == expectedValue);
+    } else if constexpr (std::is_same_v<T, VariantArray>) {
+        CHECK(v.asArray() == expectedValue);
+    } else if constexpr (std::is_same_v<T, VariantMap>) {
+        CHECK(v.asMap() == expectedValue);
+    } else if constexpr (!std::is_same_v<T, std::monostate>) {
+        FAIL("Unexpected type");
+    }
+}
+
+std::string toCbor(const Variant& v) {
+    test::Stream s;
+    REQUIRE(encodeToCBOR(v, s) == 0);
+    return s.data();
+}
+
+Variant fromCbor(const std::string& data) {
+    test::Stream s(data);
+    Variant v;
+    REQUIRE(decodeFromCBOR(v, s) == 0);
+    return v;
+}
+
+} // namespace
+
+TEST_CASE("Variant") {
+    SECTION("can be constructed from a value of one of the supported types") {
+        {
+            Variant v;
+            checkVariant<std::monostate>(v);
+        }
+        {
+            Variant v(std::monostate{});
+            checkVariant<std::monostate>(v);
+        }
+        {
+            Variant v(true);
+            checkVariant<bool>(v, true);
+        }
+        {
+            Variant v((char)123);
+            checkVariant<int>(v, 123);
+        }
+        {
+            Variant v((unsigned char)123);
+            checkVariant<int>(v, 123);
+        }
+        {
+            Variant v((short)123);
+            checkVariant<int>(v, 123);
+        }
+        {
+            Variant v((unsigned short)123);
+            checkVariant<int>(v, 123);
+        }
+        {
+            Variant v(123);
+            checkVariant<int>(v, 123);
+        }
+        {
+            Variant v(123u);
+            checkVariant<unsigned>(v, 123);
+        }
+#ifdef __LP64__
+        {
+            Variant v(123l);
+            checkVariant<int64_t>(v, 123);
+        }
+        {
+            Variant v(123ul);
+            checkVariant<uint64_t>(v, 123);
+        }
+#else
+        {
+            Variant v(123l);
+            checkVariant<int>(v, 123);
+        }
+        {
+            Variant v(123ul);
+            checkVariant<unsigned>(v, 123);
+        }
+#endif
+        {
+            Variant v(123ll);
+            checkVariant<int64_t>(v, 123);
+        }
+        {
+            Variant v(123ull);
+            checkVariant<uint64_t>(v, 123);
+        }
+        {
+            Variant v(123.0f);
```

Comment:
Check also for NaN.  I know people who use that value regularly in code to signify uninitiated values.

Reply:
Special floating point values are tested when serializing/deserializing a variant as CBOR, for example: https://github.com/particle-iot/device-os/blob/59325889b628f591756c40c9cc5fa33dc73c05a9/test/unit_tests/wiring/variant.cpp#L380-L382.